### PR TITLE
Support `autoBuild` and `deployByDefault` on Image and Kubernetes/OpenShift components 

### LINKED ISSUE #5694: implement `autoBuild` and  `deployByDefault`
devfile 2.2.0 added two new fiels we need to make sure that odo has a proper support for it

https://github.com/devfile/api/pull/709

### PR DESCRIPTION
**Pull request by rm3l**
**What type of PR is this:**
/kind feature

**What does this PR do / why we need it:**
This adds support for the following fields:
- `autoBuild` on Image Components
- `deployByDefault` on Kubernetes and OpenShift Components

See https://github.com/devfile/api/issues/852#issuecomment-1211928487

**Which issue(s) this PR fixes:**
Fixes #5694 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] [Documentation](https://deploy-preview-6654--odo-docusaurus-preview.netlify.app/docs/development/devfile#how-odo-determines-components-that-are-applied-automatically) 

**How to test changes / Special notes to the reviewer:**
Documentation is here: https://deploy-preview-6654--odo-docusaurus-preview.netlify.app/docs/development/devfile#how-odo-determines-components-that-are-applied-automatically

### CODE REVIEW

File: pkg/devfile/components.go
Comment (feloy) [reply pkg/devfile/components.go:60]: Could we already move it to libdevfile package?

Comment (rm3l) [reply pkg/devfile/components.go:60]: It would make more sense, indeed. Let me check if it could be moved there..

Comment (rm3l) [reply pkg/devfile/components.go]: Done in [`2ad44e8` (#6654)](https://github.com/redhat-developer/odo/pull/6654/commits/2ad44e8deba693a855620a6bbdfea6f724bfb475) - thanks.

File: pkg/devfile/adapters/kubernetes/component/push.go
Comment (rm3l) [reply pkg/devfile/adapters/kubernetes/component/push.go]: Done in the 3 recent commits.

**Review by feloy — APPROVED** [review]
Thanks for this work

File: pkg/libdevfile/component_kubernetes.go
Comment (valaparthvi) [inline pkg/libdevfile/component_kubernetes.go:66]: Same comment as above, I don't think this condition will ever be hit since  we are filtering the components based on their dedicated types.

File: pkg/watch/status.go
Comment (valaparthvi) [inline pkg/watch/status.go:26]:
```suggestion
	// This map allows to avoid applying them too many times upon state changes in the cluster for example.
```

File: pkg/watch/status.go
Comment (valaparthvi) [inline pkg/watch/status.go:27]: 💯

File: pkg/devfile/adapters/kubernetes/component/push.go
Comment (valaparthvi) [inline pkg/devfile/adapters/kubernetes/component/push.go:55]: nit:
I think we can print this at log level 1.
```suggestion
			klog.V(1).Infof("Skipping image component %q; already applied and not changed", c.Name)
```

File: pkg/devfile/adapters/kubernetes/component/push.go
Comment (valaparthvi) [inline pkg/devfile/adapters/kubernetes/component/push.go:68]: Would this account for any changes that might have happened to the component? I assume it is already accounted for at this point owing to [L#54](https://github.com/redhat-developer/odo/pull/6654/files#diff-014b2433b46bcae145f844d9d2e8560550acf129093f2691bc8be2097eac1a0aR54), but I'd still like to confirm.

**Review by valaparthvi — CHANGES_REQUESTED** [review]

**Comment by valaparthvi** [comment]
@rm3l The changes I've suggested are more of nit picks. Since Philippe has already approved the PR, if we want to get this PR in with this release, I am fine if you want to take them up in a followup PR.

File: tests/integration/cmd_devfile_deploy_test.go
Comment (rm3l) [reply tests/integration/cmd_devfile_deploy_test.go:678]: > On the main branch, if a component is unreferenced, we do not deploy it with `odo deploy`, this behavior is only seen on `odo dev`. However that behavior seems to have changed with this PR. I have only tested for deployByDefault=notset, but I assume it would work similarly for `deployByDefault=true`. Is this intended?
> 
Yes, that's intended. That's how I understood it, by making the rules explicitly documented and avoiding having different behaviors between the different commands.
I had added a comment related to that (it was most specifically about `odo deploy` potentially applying ServiceBinding Kubernetes components, since they are not referenced and have no `deployByDefault`): https://github.com/redhat-developer/odo/issues/5694#issuecomment-1464046208

> devfile.yaml
> ```shell
> ➜  PODMAN_CMD=echo odo deploy
>   __
>  /  \__     Running the application in Deploy mode using nodejs-prj1-api-abhz Devfile
>  \__/  \    Namespace: default
>  /  \__/    odo version: v3.8.0
>  \__/
> 
>  ⚠  You are using "default" namespace, odo may not work as expected in the default namespace.
>  ⚠  You may set a new namespace by running `odo create namespace <name>`, or set an existing one by running `odo set namespace <name>`
> 
> ↪ Deploying Kubernetes Component: my-component-unreferenced
>  ✓  Creating resource Deployment/my-component-unreferenced
>  ⚠  WARNING: Building images on Apple Silicon / M1 is not (yet) supported natively on Podman
>  ⚠  There is however a temporary workaround: https://github.com/containers/podman/discussions/12899
> 
> ↪ Building & Pushing Image: quay.io/unknown-account/myimage
>  •  Building image locally  ...
> build -t quay.io/unknown-account/myimage -f /tmp/nodejs-debug-2/Dockerfile /tmp/nodejs-debug-2
>  ✓  Building image locally [3ms]
>  •  Pushing image to container registry  ...
> push quay.io/unknown-account/myimage
>  ✓  Pushing image to container registry [2ms]
> 
> ↪ Deploying Kubernetes Component: my-component
>  ✓  Creating resource Deployment/my-component
> 
> Your Devfile has been successfully deployed
> ➜  k get deploy
> NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
> my-component                0/1     1            0           5s
> my-component-unreferenced   0/1     1            0           5s
> my-go                       0/1     1            0           11d
> ```
> 
> I wonder if the unreferenced component must be created after image build, just in case it wants to use the recently built image.

In your example, the image is referenced, so it will be applied when the `apply` command is invoked. Indeed, for resources that need to be applied automatically, this PR starts with image components first.

File: tests/integration/cmd_dev_debug_test.go
Comment (rm3l) [reply tests/integration/cmd_dev_debug_test.go:457]: Yes, that was the idea, to save some time. On the cluster, it is not needed to wait, as we'll be cleaning the entire namespace after the test. On Podman, I had some issues if not waiting, due to how Podman kind of isolates things. 
We can take a look in a different PR, if we can save more time on other tests.

**Comment by rm3l** [comment]
```

Summarizing 2 Failures:
  [FAIL] odo logs command tests when component is created and odo logs is executed when running in both Dev and Deploy mode [It] should successfully show logs of the running component
  /go/odo_1/tests/helper/helper_generic.go:79
  [FAIL] odo logs command tests when component is created and odo logs is executed when running in Dev mode [It] should successfully show logs of the running component
  /go/odo_1/tests/helper/helper_generic.go:79

Ran 439 of 800 Specs in 1251.443 seconds
FAIL! -- 437 Passed | 2 Failed | 0 Pending | 361 Skipped


Ginkgo ran 1 suite in 21m57.680223467s
```

Flaky `odo logs` tests (#6619)

/override OpenShift-Integration-tests/OpenShift-Integration-tests

File: tests/integration/cmd_devfile_deploy_test.go
Comment (valaparthvi) [reply tests/integration/cmd_devfile_deploy_test.go:678]: >Yes, that's intended. That's how I understood it, by making the rules explicitly documented and avoiding having different behaviors between the different commands.

Ah alright, I missed that comment.

>Indeed, for resources that need to be applied automatically, this PR starts with image components first.

IIUC, you mean the unreferenced k8s resources with deployByDefault:unset/true will be created after the image components are built? or the other way around?

My point was, any k8s/oc resource regardless of it's reference/deployByDefault status should be created after the image is built.

**Comment by rm3l** [comment]
/hold

I noticed that the `odo logs` tests were consistently failing. I'm investigating why..

**Comment by rm3l** [comment]
/hold cancel

The issues with `odo logs` should be fixed (well, at least, we should be back to the flaky state we see with #6619). I noticed that there were still a few places where boolean fields would be set to their default values by the Devfile parser (like in `odo init` after downloading the Devfile and parsing it), which caused some issues because `autoBuild` and `deployByDefault` were set to `false`.

**Comment by rm3l** [comment]
```
  [FAILED] Timed out after 180.000s.
  Expected process to exit.  It did not.
  In [AfterEach] at: /go/odo_1/tests/helper/helper_dev.go:200 @ 04/04/23 16:46:29.489
------------------------------
SSSâ€¢SSSSSSâ€¢â€¢â€¢Sâ€¢â€¢â€¢â€¢â€¢SSSâ€¢â€¢â€¢SSâ€¢Sâ€¢Sâ€¢â€¢Sâ€¢Sâ€¢â€¢â€¢Sâ€¢â€¢Sâ€¢â€¢SSâ€¢Sâ€¢Sâ€¢â€¢â€¢â€¢SSSâ€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢â€¢

Summarizing 1 Failure:
  [FAIL] odo dev command tests when Starting a PostgreSQL service when creating local files and dir and running odo dev - without metadata.name [AfterEach] when deleting local files and dir and waiting for sync should not list deleted dir and file in container
  /go/odo_1/tests/helper/helper_dev.go:200

Ran 439 of 800 Specs in 1531.138 seconds
FAIL! -- 438 Passed | 1 Failed | 0 Pending | 361 Skipped


Ginkgo ran 1 suite in 26m34.866445083s
```

Flaky test

/override OpenShift-Integration-tests/OpenShift-Integration-tests

**Comment by rm3l** [comment]
```

  [FAILED] Expected
      <*url.Error | 0xc000018060>: {
          Op: "Post",
          URL: "http://127.0.0.1:57428/api/newuser",
          Err: <*errors.errorString | 0xc00008a130>{s: "EOF"},
      }
  to be nil
  In [It] at: C:/Users/Administrator.ANSIBLE-TEST-VS/3615/tests/e2escenarios/e2e_test.go:306 @ 04/04/23 11:34:48.306

  There were additional failures detected.  To view them in detail run ginkgo -vv
------------------------------

Summarizing 1 Failure:
  [FAIL] E2E Test starting with non-empty Directory add Binding [It] should verify developer workflow of using binding as env in innerloop
  C:/Users/Administrator.ANSIBLE-TEST-VS/3615/tests/e2escenarios/e2e_test.go:306

Ran 7 of 7 Specs in 206.889 seconds
FAIL! -- 6 Passed | 1 Failed | 0 Pending | 0 Skipped
```

Flaky E2E test (#6582)

/override windows-integration-test/Windows-test

File: pkg/devfile/adapters/kubernetes/component/push.go
Comment (rm3l) [reply pkg/devfile/adapters/kubernetes/component/push.go:68]: Yes, changes to the component itself are taken into account with the deep comparison in [L54](https://github.com/redhat-developer/odo/pull/6654/files#diff-014b2433b46bcae145f844d9d2e8560550acf129093f2691bc8be2097eac1a0aR54).
This `devfileHasCompFn` aims at just cleaning up items from the cache to account for eventual cases where the component is removed from the Devfile.

File: tests/integration/cmd_devfile_deploy_test.go
Comment (rm3l) [reply tests/integration/cmd_devfile_deploy_test.go:678]: > > Yes, that's intended. That's how I understood it, by making the rules explicitly documented and avoiding having different behaviors between the different commands.
> 
> Ah alright, I missed that comment.
> 
> > Indeed, for resources that need to be applied automatically, this PR starts with image components first.
> 
> IIUC, you mean the unreferenced k8s resources with deployByDefault:unset/true will be created after the image components are built? or the other way around?
> 
Yes, the first way, i.e., we always start by applying image components that need to be created automatically, and then the rest.

> My point was, any k8s/oc resource regardless of it's reference/deployByDefault status should be created after the image is built.
Ah I see. That would be interesting indeed but could be a separate issue. But I think this can already be done via a composite command if you need some dependency between image and K8s/oc components, no?

File: tests/integration/cmd_dev_debug_test.go
Comment (rm3l) [reply tests/integration/cmd_dev_debug_test.go:476]: I remember I tried to do something similar using `strings.Trim{Prefix,Suffix}`, but it didn't pass because of the warning icon displayed in the beginning for this line, e.g.:
```
 ⚠  Kubernetes components are not supported on Podman. Skipping: k8s-deploybydefault-true-and-not-referenced, k8s-deploybydefault-not-set-and-not-referenced, ocp-deploybydefault-true-and-referenced, ocp-d
eploybydefault-true-and-not-referenced, ocp-deploybydefault-not-set-and-not-referenced, k8s-deploybydefault-true-and-referenced.
```

Anyways, that's a good point - I'll simplify this a little bit.

File: pkg/libdevfile/component_kubernetes.go
Comment (rm3l) [reply pkg/libdevfile/component_kubernetes.go:66]: Done in [`9544a55` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/9544a55c4c0165d95fa9a90122597fb12e2f75ed)

File: pkg/libdevfile/component_image.go
Comment (rm3l) [reply pkg/libdevfile/component_image.go:54]: Done in [`9544a55` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/9544a55c4c0165d95fa9a90122597fb12e2f75ed)

File: pkg/libdevfile/component_image.go
Comment (rm3l) [reply pkg/libdevfile/component_image.go:35]: Done in [`1a29def` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/1a29def5e45631793b8cfcd9652e79dc9f4ef476)

File: docs/website/docs/development/devfile.md
Comment (rm3l) [reply docs/website/docs/development/devfile.md:261]: Done in [`9060596` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/906059638a39b2ea6d16bae9c918b504494833a8)

File: tests/integration/cmd_dev_debug_test.go
Comment (rm3l) [reply tests/integration/cmd_dev_debug_test.go:476]: See [`a85e969` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/a85e96902d298cb0a2002e8f72f497202be1a9e2)

File: tests/integration/cmd_dev_debug_test.go
Comment (rm3l) [reply tests/integration/cmd_dev_debug_test.go:466]: See [`a85e969` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/a85e96902d298cb0a2002e8f72f497202be1a9e2)

File: tests/integration/cmd_dev_debug_test.go
Comment (rm3l) [reply tests/integration/cmd_dev_debug_test.go:420]: Done in [`be23bbf` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/be23bbfbc44aa3c8d6012c6f34a6f2196ad345f5)

File: tests/integration/cmd_dev_debug_test.go
Comment (rm3l) [reply tests/integration/cmd_dev_debug_test.go:435]: On cluster, we are using `PODMAN_CMD=echo` which is faster IMO. But yes, we could do the same thing on cluster.

File: tests/examples/source/devfiles/nodejs/devfile-autobuild-deploybydefault.yaml
Comment (rm3l) [reply tests/examples/source/devfiles/nodejs/devfile-autobuild-deploybydefault.yaml:30]: Done in [`65812b8` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/65812b857c05afc2d04c83709bf6a28bc6587c37)

**Pull request merged**
